Flaky tests  package manager update test check yum update [It] should not fail

The following test is being failing randomly in the CI:


[Fail] package manager update test check yum update [It] should not fail 
/home/jenkins/jenkins_slave/workspace/kata-containers-runtime-ubuntu-16-04-PR/go/src/github.com/kata-containers/tests/integration/docker/package_manager_test.go:68

Ran 92 of 162 Specs in 573.015 seconds
FAIL! -- 91 Passed | 1 Failed | 0 Pending | 70 Skipped --- FAIL: TestIntegration (605.89s)
FAIL

--- a/command.go
+++ b/command.go
@@ -7,6 +7,7 @@ package tests
 import (
 	"bytes"
 	"flag"
+	"fmt"
 	"os/exec"
 	"syscall"
 	"time"
@@ -52,6 +53,8 @@ func (c *Command) Run() (string, string, int) {
 func (c *Command) RunWithPipe(stdin *bytes.Buffer) (string, string, int) {
 	LogIfFail("Running command '%s %s'\n", c.cmd.Path, c.cmd.Args)
 
+	keepAliveTime := 1 * time.Minute
+
 	var stdout, stderr bytes.Buffer
 	c.cmd.Stdout = &stdout
 	c.cmd.Stderr = &stderr
@@ -72,22 +75,33 @@ func (c *Command) RunWithPipe(stdin *bytes.Buffer) (string, string, int) {
 		timeout = time.After(c.Timeout * time.Second)
 	}
 
-	select {
-	case <-timeout:
-		LogIfFail("Killing process timeout reached '%d' seconds\n", c.Timeout)
-		_ = c.cmd.Process.Kill()
-		return "", "", -1
+	keepAliveCh := time.NewTimer(keepAliveTime)
 
-	case err := <-done:
-		if err != nil {
-			LogIfFail("command failed error '%s'\n", err)
+	for {
+		select {
+		case <-timeout:
+			keepAliveCh.Stop()
+			LogIfFail("Killing process timeout reached '%d' seconds\n", c.Timeout)
+			_ = c.cmd.Process.Kill()
+			return "", "", -1
+
+		case <-keepAliveCh.C:
+			// Avoid CI (i.e jenkins) kills the process for inactivity by printing a dot
+			fmt.Println(".")
+			keepAliveCh.Reset(keepAliveTime)
+
+		case err := <-done:
+			keepAliveCh.Stop()
+			if err != nil {
+				LogIfFail("command failed error '%s'\n", err)
+			}
+
+			exitCode := c.cmd.ProcessState.Sys().(syscall.WaitStatus).ExitStatus()
+
+			LogIfFail("%+v\nTimeout: %d seconds\nExit Code: %d\nStdout: %s\nStderr: %s\n",
+				c.cmd.Args, c.Timeout, exitCode, stdout.String(), stderr.String())
+
+			return stdout.String(), stderr.String(), exitCode
 		}
-
-		exitCode := c.cmd.ProcessState.Sys().(syscall.WaitStatus).ExitStatus()
-
-		LogIfFail("%+v\nTimeout: %d seconds\nExit Code: %d\nStdout: %s\nStderr: %s\n",
-			c.cmd.Args, c.Timeout, exitCode, stdout.String(), stderr.String())
-
-		return stdout.String(), stderr.String(), exitCode
 	}
 }

--- a/integration/docker/package_manager_test.go
+++ b/integration/docker/package_manager_test.go
@@ -87,10 +87,14 @@ var _ = Describe("package manager update test", func() {
 
 	Context("check yum update", func() {
 		It("should not fail", func() {
-			Skip("Issue: https://github.com/kata-containers/tests/issues/264")
-			args = append(args, "--rm", "--name", id, CentosImage, "yum", "-y", "update")
+			args = append(args, "--rm", "-td", "--name", id, CentosImage, "sh")
 			_, _, exitCode := dockerRun(args...)
 			Expect(exitCode).To(BeZero())
+
+			exitCode = tryPackageManagerCommand(id, []string{"yum", "-y", "update"}, 0)
+			Expect(exitCode).To(BeZero())
+
+			Expect(RemoveDockerContainer(id)).To(BeTrue())
 		})
 	})
 })

--- a/integration/docker/package_manager_test.go
+++ b/integration/docker/package_manager_test.go
@@ -12,6 +12,23 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+const (
+	packageManagerTimeout  = 900
+	packageManagerMaxTries = 5
+)
+
+func tryPackageManagerCommand(container string, command []string, expectedExitCode int) int {
+	cmd := []string{container}
+	exitCode := int(-1)
+	for i := 0; i < packageManagerMaxTries; i++ {
+		_, _, exitCode = runDockerCommandWithTimeout(packageManagerTimeout, "exec", append(cmd, command...)...)
+		if exitCode == expectedExitCode {
+			break
+		}
+	}
+	return exitCode
+}
+
 var _ = Describe("package manager update test", func() {
 	var (
 		id         string

--- a/integration/docker/package_manager_test.go
+++ b/integration/docker/package_manager_test.go
@@ -69,7 +69,6 @@ var _ = Describe("package manager update test", func() {
 
 	Context("check dnf update", func() {
 		It("should not fail", func() {
-			Skip("Issue: https://github.com/clearcontainers/runtime/issues/868")
 			args = append(args, "-td", "--name", id, FedoraImage, "sh")
 			_, _, exitCode := dockerRun(args...)
 			Expect(exitCode).To(BeZero())
@@ -79,7 +78,7 @@ var _ = Describe("package manager update test", func() {
 				Expect(exitCode).To(BeZero())
 			}
 
-			_, _, exitCode = dockerExec(id, "dnf", "-y", "update")
+			exitCode = tryPackageManagerCommand(id, []string{"dnf", "-y", "update"}, 0)
 			Expect(exitCode).To(BeZero())
 
 			Expect(RemoveDockerContainer(id)).To(BeTrue())

--- a/integration/docker/package_manager_test.go
+++ b/integration/docker/package_manager_test.go
@@ -51,11 +51,19 @@ var _ = Describe("package manager update test", func() {
 		Expect(ExistDockerContainer(id)).NotTo(BeTrue())
 	})
 
-	Context("check apt-get update", func() {
+	Context("check apt-get update and upgrade", func() {
 		It("should not fail", func() {
-			args = append(args, "--rm", "--name", id, DebianImage, "apt-get", "-y", "update")
+			args = append(args, "-td", "--name", id, DebianImage, "sh")
 			_, _, exitCode := dockerRun(args...)
 			Expect(exitCode).To(BeZero())
+
+			exitCode = tryPackageManagerCommand(id, []string{"apt-get", "-y", "update"}, 0)
+			Expect(exitCode).To(BeZero())
+
+			exitCode = tryPackageManagerCommand(id, []string{"apt-get", "-y", "upgrade"}, 0)
+			Expect(exitCode).To(BeZero())
+
+			Expect(RemoveDockerContainer(id)).To(BeTrue())
 		})
 	})
 

